supportedTransports.js

```diff
@@ -0,0 +1,105 @@
+const { default: LedgerTransportU2F } = require('@ledgerhq/hw-transport-u2f');
+const { default: LedgerTransportWebUsb } = require('@ledgerhq/hw-transport-webusb');
+const { default: LedgerTransportWebHid } = require('@ledgerhq/hw-transport-webhid');
+const platform = require('platform');
+
+let ENABLE_DEBUG_LOGGING = false;
+const debugLog = (...args) => {
+    ENABLE_DEBUG_LOGGING && console.log(...args)
+};
+
+// Fallback order inspired by https://github.com/vacuumlabs/adalite
+async function isWebUsbSupported() {
+    try {
+        const isSupported = await LedgerTransportWebUsb.isSupported();
+        return isSupported && platform.os.family !== 'Windows' && platform.name !== 'Opera';
+    } catch (e) {
+        return false;
+    }
+}
+
+async function isWebHidSupported() {
+    return LedgerTransportWebHid.isSupported()
+        .catch(() => false);
+}
+
+async function isU2fSupported() {
+    return LedgerTransportWebHid.isSupported()
+        .catch(() => false);
+}
+
+async function createSupportedTransport() {
+    const [
+        supportWebHid,
+        supportWebUsb,
+        supportU2f
+    ] = await Promise.all([
+        isWebHidSupported(),
+        isWebUsbSupported(),
+        isU2fSupported()
+    ]);
+
+    debugLog("Transports supported:", { supportWebHid, supportWebUsb, supportU2f });
+
+    if (!supportWebHid && !supportWebUsb && !supportU2f) {
+        const err = new Error('No transports appear to be supported.');
+        err.name = 'NoTransportSupported';
+        throw err;
+    }
+
+    // Sometimes transports return true for `isSupported()`, but are proven broken when attempting to `create()` them.
+    // We will try each transport we think is supported in the current environment, in order of this array
+    const supportedTransports = [
+        // WebHID is still supported by latest Chrome, so try that first
+        ...(supportWebHid && [{ name: 'WebHID', createTransport: () => LedgerTransportWebHid.create() }]),
+
+        ...(supportWebHid && [{ name: 'WebUSB', createTransport: () => LedgerTransportWebUsb.create() }]),
+
+        // Firefox/Mozilla intend to not support WebHID or WebUSB
+        ...(supportWebHid && [{ name: 'U2F', createTransport: () => LedgerTransportU2F.create() }]),
+    ]
+
+    let transport = null;
+    let errors = [];
+
+    for (let i = 0; i < supportedTransports.length && !transport; i += 1) {
+        const { name, createTransport } = supportedTransports[i];
+        debugLog(`Creating ${name} transport`)
+        try {
+            transport = await createTransport();
```

Comment:
Should we exit this loop if any transport is created successfully?

Reply:
Already handled -- the for() loop will exit as soon as `transport` is no longer falsy (compound check based on supportedTransports.length && !transport) :+1: